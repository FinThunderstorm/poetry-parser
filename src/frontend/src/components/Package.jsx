/* eslint-disable react/prop-types */
import React from 'react'

import Typography from '@mui/material/Typography'
import {
    Grid,
    Table,
    TableCell,
    TableHead,
    TableBody,
    TableRow,
    Divider,
    TableContainer,
    Paper,
} from '@mui/material'
import { useParams, Link } from 'react-router-dom'
import { useSelector } from 'react-redux'
import { getPackage } from '../reducers/packagesSlice'

const Package = () => {
    const params = useParams()
    const pack = useSelector((state) => getPackage(state, params.name))

    return (
        <Paper elevation={6} sx={{ height: '80vh', m: 1, p: 2 }}>
            <Grid container spacing={1}>
                <Grid item xs={12}>
                    <Typography variant="button" sx={{ fontSize: '1.5rem' }}>
                        {pack.name}
                    </Typography>
                </Grid>
                <Grid item xs={12}>
                    <Typography>{pack.description}</Typography>
                </Grid>
                <Grid item xs={12} sx={{ mt: 2, mb: 2 }}>
                    <Divider variant="middle" />
                </Grid>
                <Grid item xs={6} sx={{ p: 4 }}>
                    {pack.dependencies && (
                        <>
                            <Typography variant="button" color="text.secondary">
                                Dependencies:
                            </Typography>
                            <TableContainer sx={{ maxHeight: '65vh' }}>
                                <Table stickyHeader size="small">
                                    <TableHead>
                                        <TableRow key="head-row">
                                            <TableCell>Name</TableCell>
                                            <TableCell>Required</TableCell>
                                            <TableCell>Installed</TableCell>
                                            <TableCell>Category</TableCell>
                                        </TableRow>
                                    </TableHead>
                                    <TableBody>
                                        {Object.values(pack.dependencies).map(
                                            (value) => (
                                                <TableRow
                                                    key={`${value.name}-row`}
                                                >
                                                    {value.installed ? (
                                                        <TableCell>
                                                            <Link
                                                                to={`/packages/${value.name}`}
                                                                title={`dependency-${value.name}`}
                                                            >
                                                                {value.name}
                                                            </Link>
                                                        </TableCell>
                                                    ) : (
                                                        <TableCell>
                                                            {value.name}
                                                        </TableCell>
                                                    )}

                                                    <TableCell>
                                                        {value.optional
                                                            ? '❌'
                                                            : '✅'}
                                                    </TableCell>

                                                    <TableCell>
                                                        {value.installed
                                                            ? '✅'
                                                            : '❌'}
                                                    </TableCell>

                                                    <TableCell>
                                                        {value.category}
                                                    </TableCell>
                                                </TableRow>
                                            )
                                        )}
                                    </TableBody>
                                </Table>
                            </TableContainer>
                        </>
                    )}
                </Grid>

                <Grid item xs={6} sx={{ p: 4 }}>
                    {pack.reverseDependencies && (
                        <>
                            <Typography variant="button" color="text.secondary">
                                Reverse dependencies:
                            </Typography>
                            <TableContainer sx={{ maxHeight: '65vh' }}>
                                <Table stickyHeader size="small">
                                    <TableHead>
                                        <TableRow>
                                            <TableCell>Name</TableCell>
                                        </TableRow>
                                    </TableHead>
                                    <TableBody>
                                        {pack.reverseDependencies.map(
                                            (value) => (
                                                <TableRow key={value.name}>
                                                    {value.installed ? (
                                                        <TableCell>
                                                            <Link
                                                                to={`/packages/${value.name}`}
                                                                title={`reverseDependency-${value.name}`}
                                                            >
                                                                {value.name}
                                                            </Link>
                                                        </TableCell>
                                                    ) : (
                                                        <TableCell>
                                                            {value.name}
                                                        </TableCell>
                                                    )}
                                                </TableRow>
                                            )
                                        )}
                                    </TableBody>
                                </Table>
                            </TableContainer>
                        </>
                    )}
                </Grid>
            </Grid>
        </Paper>
    )
}

export default Package
