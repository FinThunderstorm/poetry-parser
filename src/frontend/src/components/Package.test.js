/* eslint-disable react/prop-types */
import React from 'react'
import renderer from 'react-test-renderer'
import '@testing-library/jest-dom/extend-expect'
import { render, screen } from '@testing-library/react'
import { Provider } from 'react-redux'
import { MemoryRouter, Route, Routes } from 'react-router-dom'
import { configureStore } from '@reduxjs/toolkit'
import Package from './Package'

import packagesReducer from '../reducers/packagesSlice'

describe('Package', () => {
    const preloadedState = {
        packages: {
            packages: {
                atomicwrites: {
                    name: 'atomicwrites',
                    description: 'Atomic file writes.',
                    optional: 'false',
                    pythonVersions: [
                        '>=2.7',
                        '!=3.0.*',
                        '!=3.1.*',
                        '!=3.2.*',
                        '!=3.3.*',
                    ],
                    dependencies: {},
                    reverseDependencies: [
                        {
                            name: 'pytest',
                            installed: true,
                        },
                    ],
                },
                cachecontrol: {
                    name: 'cachecontrol',
                    description: 'httplib2 caching for requests',
                    optional: 'false',
                    pythonVersions: ['>=3.6'],
                    dependencies: {
                        lockfile: {
                            name: 'lockfile',
                            category: 'filecache',
                            optional: true,
                            source: 'extras',
                            installed: true,
                        },
                        msgpack: {
                            name: 'msgpack',
                            optional: false,
                            source: 'dependencies',
                            installed: true,
                        },
                        requests: {
                            name: 'requests',
                            optional: false,
                            source: 'dependencies',
                            installed: true,
                        },
                        redis: {
                            name: 'redis',
                            category: 'redis',
                            optional: true,
                            source: 'extras',
                            installed: false,
                        },
                    },
                    reverseDependencies: [],
                },
            },
            status: 'succeeded',
            error: null,
        },
    }

    const PackageWrapper = ({ url }) => {
        const store = configureStore({
            reducer: {
                packages: packagesReducer,
            },
            preloadedState,
        })
        return (
            <MemoryRouter initialEntries={[url]}>
                <Provider store={store}>
                    <Routes>
                        <Route path="/">
                            <Route path="packages">
                                <Route path=":name" element={<Package />} />
                            </Route>
                        </Route>
                    </Routes>
                </Provider>
            </MemoryRouter>
        )
    }

    test('Package works with reverse dependencies (atomicwrites)', () => {
        render(<PackageWrapper url="/packages/atomicwrites" />)

        const name = screen.getByText('atomicwrites', {
            exact: false,
        })
        const description = screen.getByText('Atomic file writes.', {
            exact: false,
        })
        const pytest = screen.getByText('pytest', {
            exact: false,
        })
        expect(name).toBeDefined()
        expect(description).toBeDefined()
        expect(pytest).toBeDefined()

        const tree = renderer
            .create(<PackageWrapper url="/packages/atomicwrites" />)
            .toJSON()
        expect(tree).toMatchSnapshot()
    })

    test('Package works with dependencies (cachecontrol)', () => {
        render(<PackageWrapper url="/packages/cachecontrol" />)

        const name = screen.getByText('cachecontrol', {
            exact: false,
        })
        const description = screen.getByText('httplib2 caching for requests', {
            exact: false,
        })
        const lockfile = screen.getByText('lockfile', {
            exact: false,
        })
        const filecache = screen.getByText('filecache', {
            exact: false,
        })
        const msgpack = screen.getByText('msgpack', {
            exact: false,
        })

        expect(name).toBeDefined()
        expect(description).toBeDefined()
        expect(lockfile).toBeDefined()
        expect(filecache).toBeDefined()
        expect(msgpack).toBeDefined()

        const tree = renderer
            .create(<PackageWrapper url="/packages/cachecontrol" />)
            .toJSON()
        expect(tree).toMatchSnapshot()
    })

    test('Package shows error if not found', () => {
        render(<PackageWrapper url="/packages/invalid" />)

        const error = screen.getByText('ðŸ˜­ Error while selecting package.', {
            exact: false,
        })
        expect(error).toBeDefined()

        const tree = renderer
            .create(<PackageWrapper url="/packages/invalid" />)
            .toJSON()
        expect(tree).toMatchSnapshot()
    })
})
