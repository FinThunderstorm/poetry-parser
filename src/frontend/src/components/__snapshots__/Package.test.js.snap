// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Package Package shows error if not found 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation6 css-1dxebv4-MuiPaper-root"
>
  <div
    className="css-1z0rh7t-MuiStack-root"
  >
    <span
      className="MuiTypography-root MuiTypography-button css-1sqoee-MuiTypography-root"
    >
      üò≠ Error while selecting package.
    </span>
  </div>
</div>
`;

exports[`Package Package works with dependencies (cachecontrol) 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation6 css-1dxebv4-MuiPaper-root"
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-a5rdam-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
    >
      <span
        className="MuiTypography-root MuiTypography-button css-1sqoee-MuiTypography-root"
      >
        cachecontrol
      </span>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        httplib2 caching for requests
      </p>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-eq2k9k-MuiGrid-root"
    >
      <hr
        className="MuiDivider-root MuiDivider-middle css-ft4jij-MuiDivider-root"
      />
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8 css-1u95w8b-MuiGrid-root"
    >
      <span
        className="MuiTypography-root MuiTypography-button css-1jm68wq-MuiTypography-root"
      >
        Dependencies:
      </span>
      <div
        className="MuiTableContainer-root css-e2p4vx-MuiTableContainer-root"
      >
        <table
          className="MuiTable-root MuiTable-stickyHeader css-xn82ks-MuiTable-root"
          role={null}
        >
          <thead
            className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
            role={null}
          >
            <tr
              className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
              role={null}
            >
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeSmall css-1h37f49-MuiTableCell-root"
                scope="col"
              >
                Name
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeSmall css-1h37f49-MuiTableCell-root"
                scope="col"
              >
                Required
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeSmall css-1h37f49-MuiTableCell-root"
                scope="col"
              >
                Installed
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeSmall css-1h37f49-MuiTableCell-root"
                scope="col"
              >
                Source
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeSmall css-1h37f49-MuiTableCell-root"
                scope="col"
              >
                Category
              </th>
            </tr>
          </thead>
          <tbody
            className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
            role={null}
          >
            <tr
              className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
              role={null}
            >
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
              >
                <a
                  href="/packages/lockfile"
                  onClick={[Function]}
                  title="dependency-lockfile"
                >
                  lockfile
                </a>
              </td>
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
              >
                ‚ùå
              </td>
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
              >
                ‚úÖ
              </td>
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
              >
                extras
              </td>
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
              >
                filecache
              </td>
            </tr>
            <tr
              className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
              role={null}
            >
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
              >
                <a
                  href="/packages/msgpack"
                  onClick={[Function]}
                  title="dependency-msgpack"
                >
                  msgpack
                </a>
              </td>
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
              >
                ‚úÖ
              </td>
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
              >
                ‚úÖ
              </td>
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
              >
                dependencies
              </td>
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
              />
            </tr>
            <tr
              className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
              role={null}
            >
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
              >
                <a
                  href="/packages/requests"
                  onClick={[Function]}
                  title="dependency-requests"
                >
                  requests
                </a>
              </td>
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
              >
                ‚úÖ
              </td>
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
              >
                ‚úÖ
              </td>
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
              >
                dependencies
              </td>
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
              />
            </tr>
            <tr
              className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
              role={null}
            >
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
              >
                redis
              </td>
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
              >
                ‚ùå
              </td>
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
              >
                ‚ùå
              </td>
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
              >
                extras
              </td>
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
              >
                redis
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-l2k946-MuiGrid-root"
    >
      <span
        className="MuiTypography-root MuiTypography-button css-1jm68wq-MuiTypography-root"
      >
        Reverse dependencies:
      </span>
      <div
        className="MuiTableContainer-root css-e2p4vx-MuiTableContainer-root"
      >
        <table
          className="MuiTable-root MuiTable-stickyHeader css-xn82ks-MuiTable-root"
          role={null}
        >
          <thead
            className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
            role={null}
          >
            <tr
              className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
              role={null}
            >
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeSmall css-1h37f49-MuiTableCell-root"
                scope="col"
              >
                Name
              </th>
            </tr>
          </thead>
          <tbody
            className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
            role={null}
          />
        </table>
      </div>
    </div>
  </div>
</div>
`;

exports[`Package Package works with reverse dependencies (atomicwrites) 1`] = `
<div
  className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation6 css-1dxebv4-MuiPaper-root"
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-1 css-a5rdam-MuiGrid-root"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
    >
      <span
        className="MuiTypography-root MuiTypography-button css-1sqoee-MuiTypography-root"
      >
        atomicwrites
      </span>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
    >
      <p
        className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
      >
        Atomic file writes.
      </p>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-eq2k9k-MuiGrid-root"
    >
      <hr
        className="MuiDivider-root MuiDivider-middle css-ft4jij-MuiDivider-root"
      />
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-8 css-1u95w8b-MuiGrid-root"
    >
      <span
        className="MuiTypography-root MuiTypography-button css-1jm68wq-MuiTypography-root"
      >
        Dependencies:
      </span>
      <div
        className="MuiTableContainer-root css-e2p4vx-MuiTableContainer-root"
      >
        <table
          className="MuiTable-root MuiTable-stickyHeader css-xn82ks-MuiTable-root"
          role={null}
        >
          <thead
            className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
            role={null}
          >
            <tr
              className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
              role={null}
            >
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeSmall css-1h37f49-MuiTableCell-root"
                scope="col"
              >
                Name
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeSmall css-1h37f49-MuiTableCell-root"
                scope="col"
              >
                Required
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeSmall css-1h37f49-MuiTableCell-root"
                scope="col"
              >
                Installed
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeSmall css-1h37f49-MuiTableCell-root"
                scope="col"
              >
                Source
              </th>
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeSmall css-1h37f49-MuiTableCell-root"
                scope="col"
              >
                Category
              </th>
            </tr>
          </thead>
          <tbody
            className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
            role={null}
          />
        </table>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-4 css-l2k946-MuiGrid-root"
    >
      <span
        className="MuiTypography-root MuiTypography-button css-1jm68wq-MuiTypography-root"
      >
        Reverse dependencies:
      </span>
      <div
        className="MuiTableContainer-root css-e2p4vx-MuiTableContainer-root"
      >
        <table
          className="MuiTable-root MuiTable-stickyHeader css-xn82ks-MuiTable-root"
          role={null}
        >
          <thead
            className="MuiTableHead-root css-15wwp11-MuiTableHead-root"
            role={null}
          >
            <tr
              className="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
              role={null}
            >
              <th
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-head MuiTableCell-stickyHeader MuiTableCell-sizeSmall css-1h37f49-MuiTableCell-root"
                scope="col"
              >
                Name
              </th>
            </tr>
          </thead>
          <tbody
            className="MuiTableBody-root css-apqrd9-MuiTableBody-root"
            role={null}
          >
            <tr
              className="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
              role={null}
            >
              <td
                aria-sort={null}
                className="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeSmall css-dsuxgy-MuiTableCell-root"
              >
                <a
                  href="/packages/pytest"
                  onClick={[Function]}
                  title="reverseDependency-pytest"
                >
                  pytest
                </a>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
